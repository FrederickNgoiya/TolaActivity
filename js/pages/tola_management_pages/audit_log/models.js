import { observable, computed, action, runInAction } from "mobx"
import api from './api'

export class ProgramAuditLogStore {
    @observable program_id = null
    @observable log_rows = []
    @observable fetching = false
    @observable current_page = 0
    @observable details_target = null

    constructor(program_id) {
        this.program_id = program_id
        this.fetchProgramAuditLog()
    }

    @action
    fetchProgramAuditLog() {
        this.fetching = true

        api.fetchProgramAuditLogWithFilter(this.program_id, this.current_page + 1).then(results => {
            runInAction(() => {
                this.fetching = false
                this.log_rows = results.logs
            })
        })
    }

    @action
    toggleDetailsTarget(row_id) {
        if(this.details_target == row_id) {
            this.details_target = null
        } else {
            this.details_target = row_id
        }
    }

}
