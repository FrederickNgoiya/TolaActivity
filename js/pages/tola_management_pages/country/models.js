import { observable, computed, action, runInAction } from "mobx";


const new_objective_data = {
    id: 'new',
    name: '',
    description: '',
    status: '',
}


export class CountryStore {

    //filter options
    @observable organizations = {}
    @observable users = []
    @observable sectors = []

    @observable filters = {
        countries: [],
        organizations: [],
        sectors: [],
        programStatus: null,
        programs: [],
    }

    @observable allCountries = []
    @observable countries = []
    @observable country_count = 0
    @observable new_country = null
    @observable fetching_main_listing = false
    @observable current_page = 0
    @observable total_pages = null
    @observable bulk_targets = new Map()
    @observable bulk_targets_all = false

    @observable editing_target = null
    @observable editing_errors = {}
    @observable fetching_editing_data = true
    @observable editing_objectives_data = []
    @observable editing_objectives_errors = {}
    @observable editing_disaggregations_data = []
    @observable editing_disaggregations_errors = {}
    @observable saving = false

    @observable bulk_targets = new Map()
    @observable applying_bulk_updates = false
    @observable bulk_targets_all = false

    constructor(
        api,
        initialData,
    ) {
        this
        this.api = api
        Object.assign(this, initialData)
        this.fetchCountries()
    }

    marshalFilters(filters) {
        return Object.entries(filters).reduce((xs, [filterKey, filterValue]) => {
            if (Array.isArray(filterValue)) {
                xs[filterKey] = filterValue.map(x => x.value)
            } else if (filterValue) {
                xs[filterKey] = filterValue.value
            }
            return xs
        }, {})
    }

    @action
    fetchCountries() {
        this.fetching_main_listing = true
        this.api.fetchCountries(this.current_page + 1, this.marshalFilters(this.filters)).then(results => {
            runInAction(() => {
                this.fetching_main_listing = false
                this.countries = results.results
                this.country_count = results.total_results
                this.total_pages = results.total_pages
                this.next_page =results.next_page
                this.previous_page = results.previous_page
            })
        })

    }

    @action
    applyFilters() {
        this.current_page = 0
        this.fetchCountries()
    }

    @action
    changePage(page) {
        if (page.selected == this.current_page) {
            return
        }
        this.current_page = page.selected
        this.bulk_targets = new Map()
        this.bulk_targets_all = false;
        this.fetchCountries()
    }

    @action
    changeFilter(filterKey, value) {
        this.filters = Object.assign(this.filters, {[filterKey]: value})
    }

    @action
    clearFilters() {
        let clearFilters = {
            countries: [],
            organizations: [],
            sectors: [],
            programStatus: null,
            programs: [],
        }
        this.filters = Object.assign(this.filters, clearFilters);
    }

    @action
    toggleEditingTarget(id) {
        if(this.editing_target == 'new') {
            this.countries.shift()
            this.editing_errors = {}
        }

        if(this.editing_target == id) {
            this.editing_target = false
            this.editing_errors = {}
        } else {
            this.editing_target = id
            this.fetching_editing_data = true
            Promise.all([
                this.api.fetchCountryObjectives(id),
                this.api.fetchCountryDisaggregations(id),
            ]).then(([objectives_resp, disaggregations_resp]) => {
                runInAction(() => {
                    this.fetching_editing_data = false
                    this.editing_objectives_data = objectives_resp.data
                    this.editing_disaggregations_data = disaggregations_resp.data
                })
            })
        }
    }

    updateLocalList(updated) {
        this.countries = this.countries.reduce((acc, current) => {
            if (current.id == updated.id) {
                acc.push(updated)
            } else {
                acc.push(current)
            }
            return acc
        }, [])
    }

    onSaveSuccessHandler() {
        PNotify.success({text: gettext("Successfully Saved"), delay: 5000})
    }

    onSaveErrorHandler(message) {
        PNotify.error({text: message || gettext("Saving Failed"), delay: 5000})
    }

    onDeleteSuccessHandler() {
        PNotify.success({text: gettext("Successfully Deleted"), delay: 5000})
    }

    @action
    addCountry() {
        if(this.editing_target == 'new') {
            this.countries.shift()
        }

        let new_country_data = {
            id: "new",
            country: "",
            description: "",
            code: "",
            organizations: [],
        }
        this.countries.unshift(new_country_data)
        this.editing_target = 'new'
    }

    @action
    saveNewCountry(country_data) {
        country_data.id = null
        this.saving = true
        this.api.createCountry(country_data).then(response => {
            runInAction(()=> {
                this.saving = false
                this.editing_target = false
                this.countries.shift()
                this.countries.unshift(response.data)
            })
        }).catch(errors => {
            runInAction(()=> {
                this.saving = false
                this.editing_errors = errors.response.data
                this.onSaveErrorHandler(errors.response.data.detail)
            })
        })
    }

    @action updateCountry(id, country_data) {
        this.saving = true
        this.api.updateCountry(id, country_data).then(response => {
            runInAction(() => {
                this.saving = false
                this.editing_target = false
                this.updateLocalList(response.data)
                this.onSaveSuccessHandler()
            })
        }).catch((errors) => {
            runInAction(() => {
                this.saving = false
                this.editing_errors = errors.response.data
                this.onSaveErrorHandler(errors.response.data.detail)
            })
        })
    }

    @action addObjective() {
        if (this.editing_objectives_data.find(objective => objective.id=='new')) {
            return
        }
        this.editing_objectives_data = [...this.editing_objectives_data, new_objective_data]
    }

    @action updateObjective(id, data) {
        this.editing_objectives_errors = {}
        this.api.updateObjective(id, data).then(response => {
            runInAction(() => {
                this.onSaveSuccessHandler()
                let updatedObjective = response.data
                this.editing_objectives_data = this.editing_objectives_data.map(objective => {
                    if (objective.id == updatedObjective.id) {
                        return updatedObjective
                    }
                    return objective
                })
            })
        }).catch((errors) => {
            runInAction(() => {
                this.saving = false
                this.editing_objectives_errors = errors.response.data
                this.onSaveErrorHandler(errors.response.data.detail)
            })
        })
    }

    @action createObjective(data) {
        this.editing_objectives_errors = {}
        this.api.createObjective(data).then(response => {
            runInAction(() => {
                this.onSaveSuccessHandler()
                let newObjective = response.data
                this.editing_objectives_data = [...this.editing_objectives_data.filter(objective => objective.id!='new'), newObjective]
            })
        }).catch((errors) => {
            runInAction(() => {
                this.saving = false
                this.editing_objectives_errors = errors.response.data
                this.onSaveErrorHandler(errors.response.data.detail)
            })
        })
    }

    @action deleteObjective(id) {
        if (id=='new') {
            this.editing_objectives_data = this.editing_objectives_data.filter(objective => objective.id!='new')
            return
        }
        this.api.deleteObjective(id).then(response => {
            runInAction(() => {
                this.editing_objectives_data = this.editing_objectives_data.filter(objective => objective.id!=id)
                this.onDeleteSuccessHandler()
            })
        }).catch((errors) => {
            runInAction(() => {
                this.onSaveErrorHandler(errors.response.data.detail)
            })
        })
    }

    @action clearObjectiveEditingErrors() {
        this.editing_objectives_errors = {}
    }

    @action clearDisaggregationEditingErrors() {
        this.editing_disaggregations_errors = {}
    }

    @action addDisaggregation() {
        const new_disaggregation_data = {
            id: 'new',
            disaggregation_type: "",
            labels: [],
        }
        if (this.editing_disaggregations_data.find(disaggregation => disaggregation.id=='new')) {
            return
        }
        this.editing_disaggregations_data = [...this.editing_disaggregations_data, new_disaggregation_data]
    }

    @action deleteDisaggregation(id) {
        if (id=='new') {
            this.editing_disaggregations_data = this.editing_disaggregations_data.filter(disagg=>disagg.id!='new')
            return
        }
        /*
        this.api.deleteDisaggregation(id).then(response => {
            runInAction(() => {
                this.editing_disaggregations_data = this.editing_disaggregations_data.filter(disagg => disagg.id!=id)
                this.onDeleteSuccessHandler()
            })
        }
        */
    }

    @action updateDisaggregation(id, data) {
        this.editing_disaggregations_errors = {}
        this.api.updateDisaggregation(id, data).then(response => {
            runInAction(() => {
                this.onSaveSuccessHandler()
                let updatedDisaggregation = response.data
                this.editing_disaggregations_data = this.editing_disaggregations_data.map(disaggregation => {
                    if (disaggregation.id == updatedDisaggregation.id) {
                        return updatedDisaggregation
                    }
                    return disaggregation
                })
            })
        }).catch((errors) => {
            this.saving = false
            this.editing_disaggregations_errors = errors.response.data
            this.onSaveErrorHandler()
        })
    }

    @action createDisaggregation(data) {
        this.editing_disaggregations_errors = {}
        this.api.createDisaggregation(data).then(response => {
            runInAction(() => {
                this.onSaveSuccessHandler()
                const newDisaggregation = response.data
                this.editing_disaggregations_data = [...this.editing_disaggregations_data.filter(disaggregation => disaggregation.id!='new'), newDisaggregation]
            })
        }).catch((errors) => {
            runInAction(() => {
                this.saving = false
                this.editing_disaggregations_errors = errors.response.data
                this.onSaveErrorHandler()
            })
        })
    }

}
