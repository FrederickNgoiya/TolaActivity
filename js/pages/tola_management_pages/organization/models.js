import { observable, computed, action, runInAction } from "mobx";
import api from './api';

const default_organization = {
    id: null ,
    is_active: false,
    mode_of_contact: "",
    name: "",
    organization_url: null,
    primary_address: "",
    primary_contact_email: "",
    primary_contact_name: "",
    primary_contact_phone: "",
    sectors: [],
}

export class OrganizationStore {
    @observable organizations = {}
    @observable organizations_listing = []
    @observable organizations_count = 0
    @observable total_pagees = 0
    @observable fetching = false
    @observable fetching_editing_target = false
    @observable current_page = 0
    @observable saving = false

    @observable bulk_targets = new Map()
    @observable bulk_targets_all = false

    available_programs = {}
    available_organizations = {}
    available_sectors = {}
    available_countries = {}
    program_selections = []
    organization_selections = []
    sector_selections = []
    country_selections = []

    @observable editing_target = null
    @observable editing_target_data = {...default_organization}
    @observable editing_target_history = []
    @observable editing_errors = {}

    @observable filters = {
        countries: [],
        organizations: [],
        programs: [],
        sectors: [],
        organization_status: '',
    }

    organization_status_options = [
        {value: 1, label: 'Active'},
        {value: 0, label: 'Inactive'}
    ]

    constructor(programs, organizations, sectors, countries, country_filter, program_filter) {
        this.available_programs = programs
        this.available_organizations = organizations
        this.available_sectors = sectors
        this.available_countries = countries
        this.organization_selections = Object.entries(organizations).map(([id, org]) => ({value: org.id, label: org.name}))
        this.program_selections = Object.entries(programs).map(([id, program]) => ({value: program.id, label: program.name}))
        this.sector_selections = Object.entries(sectors).map(([id, sector]) => ({value: sector.id, label: sector.name}))
        this.country_selections = Object.entries(countries).map(([id, country]) => ({value: country.id, label: country.name}))
        this.filters.countries = country_filter.map(id => this.available_countries[id]).map(country => ({label: country.name, value: country.id}))
        this.filters.programs = program_filter.filter(id => programs[id]).map(id => ({label: programs[id].name, value: id}))
        this.fetchOrganizations()
    }

    marshalFilters(filters) {
        return Object.entries(filters).reduce((xs, x) => {
            if(Array.isArray(x[1])) {
                xs[x[0]] = x[1].map(x => x.value)
            } else {
                xs[x[0]] = x[1].value
            }
            return xs
        }, {})
    }

    updateLocalOrganization(id, applied_data, aggregates) {
        this.organizations[id] = {
            id: id,
            name: applied_data.name,
            program_count: aggregates.program_count,
            user_count: aggregates.user_count,
            is_active: applied_data.is_active
        }
    }

    onSaveErrorHandler() {
        PNotify.error({text: 'Saving Failed', delay: 5000});
    }

    onSaveSuccessHandler() {
        PNotify.success({text: 'Successfully Saved', delay: 5000})
    }

    @action
    fetchOrganizations() {
        this.fetching = true

        api.fetchOrganizationsWithFilter(this.current_page + 1, this.marshalFilters(this.filters)).then(results => {
            runInAction(() => {
                this.fetching = false
                this.organizations = results.organizations.reduce((xs, x) => {
                    xs[x.id] = x
                    return xs
                }, {})
                this.organizations_listing = results.organizations.map(o => o.id)
                this.organizations_count = results.total_organizations
                this.total_pages = results.total_pages
                this.bulk_targets = new Map(Object.entries(this.organizations).map(([_, organization]) => [organization.id, false]))
            })
        })
    }

    @action
    createOrganization() {
        const new_organization = {
            id: "new",
            name: "",
            program_count: 0,
            user_count: 0,
            is_active: false
        }
        if(this.editing_target !== "new") {
            this.organizations_listing.unshift("new")
        }
        this.organizations["new"] = new_organization
        this.editing_target = new_organization.id
        this.editing_target_data = {...default_organization}
    }

    @action
    updateOrganizationProfile(id, new_data) {
        this.saving = true
        api.updateOrganization(id, new_data).then(updated_data => api.fetchOrganizationAggregates(id).then(aggregates => {
            runInAction(() => {
                this.saving = false
                this.updateLocalOrganization(id, updated_data, aggregates)
                this.editing_target = null
                this.editing_target_data = {...default_organization}
            })
            this.onSaveSuccessHandler()
        })).catch((errors) => {
            runInAction(() => {
                this.saving = false
                this.editing_errors = errors.response.data
            })
            this.onSaveErrorHandler()
        })
    }

    @action
    saveNewOrganization(new_data) {
        this.saving = true
        new_data.is_active = true;
        api.createOrganization(new_data).then(result => {
            runInAction(() => {
                this.saving = false
                this.updateLocalOrganization(result.id, result, {program_count: 0, user_count: 0})
                this.organizations_listing.shift()
                delete this.organizations["new"]
                this.organizations_listing.unshift(result.id)
                this.editing_target = null
                this.editing_target_data = {...default_organization}
                this.bulk_targets = new Map(Object.entries(this.organizations).map(([_, organization]) => [organization.id, false]))
            })
            this.onSaveSuccessHandler()
        }).catch(error => {
            runInAction(() => {
                this.saving = false
                this.editing_errors = errors.response.data
            })
            this.onSaveErrorHandler()
        })
    }

    @action
    saveNewOrganizationAndAddAnother(new_data) {
        this.saving = true
        new_data.is_active = true;
        api.createOrganization(new_data).then(result => {
            runInAction(() => {
                this.saving = false
                this.updateLocalOrganization(result.id, result, {program_count: 0, user_count: 0})
                this.organizations_listing.shift()
                delete this.organizations["new"]
                this.organizations_listing.unshift(result.id)
                this.editing_target = null
                this.editing_target_data = {...default_organization}
                this.bulk_targets = new Map(Object.entries(this.organizations).map(([_, organization]) => [organization.id, false]))
            })
            this.onSaveSuccessHandler()
        }).catch(error => {
            runInAction(() => {
                this.saving = false
                this.editing_errors = errors.response.data
            })
            this.onSaveErrorHandler()
        })
    }

    @action
    changeSectorFilter(sectors) {
        this.filters.sectors = sectors
    }

    @action
    changeCountryFilter(countries) {
        this.filters.countries = countries
    }

    @action
    changeProgramFilter(programs) {
        this.filters.programs = programs
    }

    @action
    changeOrganizationFilter(organizations) {
        this.filters.organizations = organizations
    }

    @action
    changeOrganizationStatusFilter(status) {
        this.filters.organization_status = status
    }

    @action
    changePage(page) {
        if(this.current_page != page.selected) {
            this.current_page = page.selected
            this.fetchOrganizations()
        }
    }

    @action
    toggleBulkTargetsAll() {
        this.bulk_targets_all = !this.bulk_targets_all;
        if(this.bulk_targets_all) {
            this.bulk_targets.forEach((val, key, map) => {
                map.set(key, true)
            })
        } else {
            this.bulk_targets.forEach((val, key, map) => {
                map.set(key, false)
            })
        }
    }

    @action
    toggleEditingTarget(organization_id) {
        this.editing_target_data = {...default_organization}

        if(this.editing_target == "new") {
            this.organizations_listing.shift()
        }

        if(this.editing_target == organization_id) {
            this.editing_target = false
        } else {
            this.editing_target = organization_id
            this.fetching_editing_target = true
            if(!(this.editing_target == 'new')) {
                Promise.all([api.fetchOrganization(organization_id), api.fetchOrganizationHistory(organization_id)]).then(([organization, history]) => {
                    runInAction(() => {
                        this.fetching_editing_target = false
                        this.editing_target_data = organization
                        this.editing_target_history = history
                    })
                })
            }
        }
    }

    @action
    toggleBulkTarget(target_id) {
        this.bulk_targets.set(target_id, !this.bulk_targets.get(target_id))
    }

    @action
    clearFilters() {
        this.filters = {
            countries: [],
            organizations: [],
            programs: [],
            sectors: [],
            organization_status: '',
        }
    }
}
