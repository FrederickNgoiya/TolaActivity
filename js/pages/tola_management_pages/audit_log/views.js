import React from 'react';
import { observer } from "mobx-react"
import Select from 'react-select'
import classNames from 'classnames'
import ManagementTable from 'components/management-table'
import Pagination from 'components/pagination'
import CheckboxedMultiSelect from 'components/checkboxed-multi-select'
import Expander from 'components/expander'

import LoadingSpinner from 'components/loading-spinner'

const ResultChangeset = ({data, name, pretty_name}) => {
    if(name == 'evidence_url') {
        return <div className="change__field"><strong>{pretty_name}</strong>: {(data != 'N/A')?<a href={data}>Link</a>:data}</div>
    } else {
        return <div className="change__field"><strong>{pretty_name}</strong>: {data}</div>
    }
}

const ProgramDatesChangeset = ({data, name, pretty_name}) => {
    return <p>{pretty_name}: {data}</p>
}

const IndicatorChangeset = ({data, name, pretty_name}) => {
    if(name == 'targets') {
        return <div className="changelog__change__targets">
            <h4 className="text-small">{gettext('Targets changed')}</h4>
            {Object.entries(data).map(([id, target]) => {
                return <div className="change__field" key={id}><strong>{target.name}:</strong> {target.value}</div>
            })}
        </div>
    } else {
        return <div className="change__field">
            <strong>{pretty_name}:</strong> {(data !== null && data !== undefined)?data.toString():gettext('N/A')}
        </div>
    }
}

class ChangesetEntry extends React.Component {
    renderType(type, data, name, pretty_name) {
        switch(type) {
            case 'indicator_changed':
            case 'indicator_created':
            case 'indicator_deleted':
                return <IndicatorChangeset data={data} name={name} pretty_name={pretty_name} />
                break
            case 'result_changed':
            case 'result_created':
            case 'result_deleted':
                return <ResultChangeset data={data} name={name} pretty_name={pretty_name} />
                break
            case 'program_dates_changed':
                return <ProgramDatesChangeset data={data} name={name} pretty_name={pretty_name} />
                break
        }
    }

    render() {
        const {data, type, name, pretty_name} = this.props
        return this.renderType(type, data, name, pretty_name)
    }
}

export const IndexView = observer(
    ({store}) => {
        return <div id="audit-log-index-view">
            <header class="page-title">
                <h1><small>{gettext("Indicator change log:")}</small> {store.program_name}</h1>
            </header>

            <div className="admin-list__controls">
                <div className="controls__bulk-actions" />
                <div className="controls__buttons">
                    <a className="btn btn-secondary btn-sm" href={`/api/tola_management/program/${store.program_id}/export_audit_log`}>
                        <i className="fas fa-download"></i>
                        {gettext("Excel")}
                    </a>
                </div>
            </div>
            <div className="admin-list__table">
                <LoadingSpinner isLoading={store.fetching}>
                    <table className="table table-sm table-bordered bg-white text-small changelog">
                        <thead>
                            <tr>
                                <th className="text-nowrap">{gettext("Date and Time")}</th>
                                <th className="text-nowrap">{gettext("No.")}</th>
                                <th className="text-nowrap">{gettext("Indicator")}</th>
                                <th className="text-nowrap">{gettext("User")}</th>
                                <th className="text-nowrap">{gettext("Organization")}</th>
                                <th className="text-nowrap">{gettext("Change Type")}</th>
                                <th className="text-nowrap">{gettext("Previous Entry")}</th>
                                <th className="text-nowrap">{gettext("New Entry")}</th>
                                <th className="text-nowrap">{gettext("Rationale")}</th>
                            </tr>
                        </thead>
                        {store.log_rows.map(data => <tbody>
                            <tr className="changelog__entry__header is-expanded">
                                <td>{data.date}</td>
                                <td>{(data.indicator)?data.indicator.number:gettext('N/A')}</td>
                                <td>{(data.indicator)?data.indicator.name:gettext('N/A')}</td>
                                <td>{data.user}</td>
                                <td>{data.organization}</td>
                                <td>{data.pretty_change_type}</td>{/* SWEET FANCY MOSES WHAT IS THIS */}
                                <td></td>
                                <td></td>
                                <td></td>
                            </tr>
                            <tr className="changelog__entry__row" key={data.id}>
                                <td></td>
                                <td></td>
                                <td></td>
                                <td></td>
                                <td></td>
                                <td></td>
                                <td className="changelog__change--prev">
                                    {data.diff_list.map(changeset => {
                                        return <ChangesetEntry key={changeset.name} name={changeset.name} pretty_name={changeset.pretty_name} type={data.change_type} data={changeset.prev} />
                                    })}
                                </td>
                                <td className="changelog__change--new">
                                    {data.diff_list.map(changeset => {
                                         return <ChangesetEntry key={changeset.name} name={changeset.name} pretty_name={changeset.pretty_name} type={data.change_type} data={changeset.new} />
                                    })}
                                </td>
                                <td className="changelog__change--rationale">{data.rationale}</td>
                            </tr>
                        </tbody>
                        )}
                    </table>
                </LoadingSpinner>
                <div className="admin-list__metadata">
                    <div className="metadata__count text-muted text-small">{store.entries_count?`${store.entries_count} ${gettext("entries")}`:`--`}</div>
                    <div className="metadata__controls">
                        {store.total_pages &&
                         <Pagination
                             pageCount={store.total_pages}
                             initialPage={store.current_page}
                             onPageChange={page => store.changePage(page)} />
                        }
                    </div>
                </div>
            </div>
        </div>
    }
)
