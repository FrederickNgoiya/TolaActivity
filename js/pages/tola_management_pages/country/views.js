import React from 'react'
import { observer } from "mobx-react"
import CheckboxedMultiSelect from 'components/checkboxed-multi-select'
import ManagementTable from 'components/management-table'
import Pagination from 'components/pagination'
import CountryEditor from './components/country_editor'
import EditCountryProfile from './components/edit_country_profile'
import EditDisaggregations from './components/edit_disaggregations'
import EditObjectives from './components/edit_objectives'
import LoadingSpinner from 'components/loading-spinner'
import FoldingSidebar from 'components/folding-sidebar'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

const CountryFilter = observer(({store, filterOptions}) => {
    return <div className="form-group">
        <label htmlFor="countries_filter">{gettext("Countries")}</label>
        <CheckboxedMultiSelect
            value={store.filters.countries}
            options={filterOptions}
            onChange={(e) => store.changeFilter('countries', e)}
            placeholder={gettext("None Selected")}
            id="countries_filter" />
    </div>
})

const OrganizationFilter = observer(({store, filterOptions}) => {
    return <div className="form-group">
        <label htmlFor="organizations_filter">{gettext("Organizations")}</label>
        <CheckboxedMultiSelect
            value={store.filters.organizations}
            options={filterOptions}
            onChange={(e) => store.changeFilter('organizations', e)}
            placeholder={gettext("None Selected")}
            id="organizations_filter" />
    </div>
})

const ProgramFilter = observer(({store, filterOptions}) => {
    return <div className="form-group">
        <label htmlFor="programs-filter">{gettext("Programs")}</label>
        <CheckboxedMultiSelect
            value={store.filters.programs}
            options={filterOptions}
            onChange={(e) => store.changeFilter('programs', e)}
            placeholder={gettext("None Selected")}
            id="programs-filter" />
    </div>
})


export const IndexView = observer(
    ({store}) => {
        const countryFilterOptions = store.allCountries.map(country => {return {value: country.id, label: country.country}})
        const organizationFilterOptions = Object.entries(store.organizations).map(([id, org]) => ({value: org.id, label: org.name}))
        const programFilterOptions = Object.entries(store.allPrograms).map(([id, program]) => ({value: program.id, label: program.name}))

        return <div id="country-management-index-view" className="row">
            <FoldingSidebar>
                <div className="filter-section">
                    <OrganizationFilter store={store} filterOptions={organizationFilterOptions} />
                    <ProgramFilter store={store} filterOptions={programFilterOptions} />
                    <CountryFilter store={store} filterOptions={countryFilterOptions} />
                    <div className="filter-buttons">
                        <button className="btn btn-primary" onClick={() => store.applyFilters()}>{gettext("Apply")}</button>
                        <button className="btn btn-inverse" onClick={() => store.clearFilters()}>{gettext("Reset")}</button>
                    </div>
                </div>
            </FoldingSidebar>
            <div className="col admin-list">
                <div className="admin-list__controls">
                {store.is_superuser &&
                <div>
                    <a href="#" tabIndex="0" className="btn btn-link btn-add" onClick={() => store.addCountry()}>
                        <FontAwesomeIcon icon={'plus-circle'} />
                        {gettext("Add Country")}
                    </a>
                </div>
                }
                </div>
                <LoadingSpinner isLoading={store.fetching_main_listing || store.applying_bulk_updates }>
                    <div className="admin-list__table">
                        <ManagementTable
                            newData={store.new_country}
                            data={store.countries}
                            keyField="id"
                            HeaderRow={({Col, Row}) =>
                                <Row>
                                    <Col size=".2"></Col>
                                    <Col size="2" class="td--stretch">{gettext("Country")}</Col>
                                    <Col>{gettext("Organizations")}</Col>
                                    <Col>{gettext("Programs")}</Col>
                                    <Col>{gettext("Users")}</Col>
                                </Row>
                            }
                            Row={({Col, Row, data}) =>
                            <Row
                                expanded={data.id == store.editing_target}
                                Expando={({Wrapper}) =>
                                    <Wrapper>
                                        <CountryEditor
                                            new={data.id == 'new'}
                                            ProfileSection={observer(() =>
                                                <EditCountryProfile
                                                    new={data.id == 'new'}
                                                    country_data={data}
                                                    organizationOptions={organizationFilterOptions}
                                                    onUpdate={(id, data) => store.updateCountry(id, data)}
                                                    onCreate={(new_country_data) => store.saveNewCountry(new_country_data)}
                                                    errors={store.editing_errors}
                                                />)}
                                            StrategicObjectiveSection={observer(() =>
                                                <LoadingSpinner isLoading={store.fetching_editing_data}>
                                                    <EditObjectives
                                                        country_id={data.id}
                                                        objectives={store.editing_objectives_data}
                                                        addObjective={() => store.addObjective()}
                                                        onUpdate={(id, data) => store.updateObjective(id, data)}
                                                        onCreate={(data) => store.createObjective(data)}
                                                        onDelete={(id) => store.deleteObjective(id)}
                                                        errors={store.editing_objectives_errors}
                                                        clearErrors={() => store.clearObjectiveEditingErrors()}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            DisaggregationSection={observer(() =>
                                                <LoadingSpinner isLoading={store.fetching_editing_data}>
                                                    <EditDisaggregations
                                                        country_id={data.id}
                                                        disaggregations={store.editing_disaggregations_data}
                                                        addDisaggregation={() => store.addDisaggregation()}
                                                        onDelete={(id) => store.deleteDisaggregation(id)}
                                                        onUpdate={(id, data) => store.updateDisaggregation(id, data)}
                                                        onCreate={(data) => store.createDisaggregation(data)}
                                                        errors={store.editing_disaggregations_errors}
                                                        clearErrors={() => store.clearDisaggregationEditingErrors()}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            fetchObjectives={(countryId) => store.fetchObjectives(countryId)}
                                        />
                                    </Wrapper>
                                }>
                                    <Col size="0.2">
                                    </Col>
                                    <Col size="2" className="td--stretch">
                                        <div className="icon__clickable" onClick={() => store.toggleEditingTarget(data.id)} >
                                            <FontAwesomeIcon icon={'globe'} />&nbsp;
                                            {data.country || "---"}
                                        </div>
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.organizations.length ?
                                            <a href={`/tola_management/organization/?countries[]=${data.id}`}>
                                                <FontAwesomeIcon icon={'building'} />&nbsp;
                                                {data.organizations.length} {gettext("Organizations")}
                                            </a>
                                        : '---'}
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.programCount ?
                                            <a href={`/tola_management/program/?countries[]=${data.id}`}>
                                                <FontAwesomeIcon icon={'cubes'} />&nbsp;
                                                {data.programCount} {gettext("Programs")}
                                            </a>
                                        : "---"}
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.user_count ?
                                            <a href={`/tola_management/user/?countries[]=${data.id}`}>
                                                <FontAwesomeIcon icon={'users'} />&nbsp;
                                                {data.user_count} {gettext("Users")}
                                            </a>
                                        : '---'  }
                                    </Col>
                                </Row>
                            }
                        />
                    </div>
                </LoadingSpinner>
                <div className="admin-list__metadata">
                    <div id="users-count">{store.country_count ? `${store.country_count} ${gettext("countries")}`:`---`}</div>
                    <div id ="pagination-controls">
                        {store.total_pages &&
                         <Pagination
                            pageCount={store.total_pages}
                            initialPage={store.current_page}
                            onPageChange={page => store.changePage(page)} />
                        }
                    </div>
                </div>
            </div>
        </div>
    }
)
