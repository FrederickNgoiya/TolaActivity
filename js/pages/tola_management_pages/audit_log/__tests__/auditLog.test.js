import React from 'react';
import { resultLogChanged, resultLogDisaggs } from '../__fixtures__/resultLogFixtures';
import { ProgramAuditLogStore } from "../models"
import { ResultChangeset, DisaggregationDiffs } from "../views"
import renderer from 'react-test-renderer';

describe("Audit log", () => {

    it("orders disaggregations correctly", () => {
        let disaggList = Object.values(resultLogDisaggs);
        let component = renderer.create(
            <DisaggregationDiffs
                disagg_type={disaggList['1'].type}
                disagg_diffs={disaggList}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();

        resultLogDisaggs["10"]["custom_sort"] = 4;
        resultLogDisaggs["20"]["custom_sort"] = 1;
        disaggList = Object.values(resultLogDisaggs);
        component = renderer.create(
            <DisaggregationDiffs
                disagg_type={disaggList['1'].type}
                disagg_diffs={disaggList}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("orders disaggregations types correctly", () => {
        const component = renderer.create(
            <ResultChangeset
                name={resultLogChanged.name}
                pretty_name={resultLogChanged.pretty_name}
                data={resultLogChanged.new}
            />
        );
        expect(component).toMatchSnapshot();
    })
});
