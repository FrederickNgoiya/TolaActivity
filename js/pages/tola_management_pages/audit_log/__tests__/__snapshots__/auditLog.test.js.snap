// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit log orders disaggregations correctly 1`] = `
<div>
  <p
    className="disagg-type__title"
  >
    Type of Leader
  </p>
  <div
    className="change__field"
  >
    <strong>
      Value One
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value Two
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value Three
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value Four
      :
    </strong>
     
    
  </div>
</div>
`;

exports[`Audit log orders disaggregations correctly 2`] = `
<div>
  <p
    className="disagg-type__title"
  >
    Type of Leader
  </p>
  <div
    className="change__field"
  >
    <strong>
      Value Four
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value Two
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value Three
      :
    </strong>
     
    
  </div>
  <div
    className="change__field"
  >
    <strong>
      Value One
      :
    </strong>
     
    
  </div>
</div>
`;

exports[`Audit log orders disaggregations types correctly 1`] = `
<div
  className="changelog__change__targets"
>
  <h4
    className="text-small"
  >
    Disaggregated values
  </h4>
  <div>
    <p
      className="disagg-type__title"
    >
      Another Disaggregation
    </p>
    <div
      className="change__field"
    >
      <strong>
        Type Two Value One
        :
      </strong>
       
      18.00
    </div>
  </div>
  <div>
    <p
      className="disagg-type__title"
    >
      Type of Leader
    </p>
    <div
      className="change__field"
    >
      <strong>
        Type One Value Two
        :
      </strong>
       
      23.00
    </div>
    <div
      className="change__field"
    >
      <strong>
        Type One Value One
        :
      </strong>
       
      2.00
    </div>
  </div>
</div>
`;
